ground/src/main/java/com/google/android/ground/system/LocationManager.kt

```diff
@@ -79,6 +86,40 @@ constructor(
       .onErrorReturn { Result.failure(it) }
   }
 
+  fun toggleLocationLock() {
+    if (locationLockState.value.getOrDefault(false)) {
+      unlock()
+
+      disableLocationUpdates()
+    } else {
+      permissionsManager
+        .obtainPermission(permission.ACCESS_FINE_LOCATION)
+        .andThen(
+          completeIf {
+            lock()
+            true
+          }
+        )
+        .andThen(settingsManager.enableLocationSettings(FINE_LOCATION_UPDATES_REQUEST))
+        .andThen(requestLocationUpdates())
+        .toSingle {
+          lock()
+          Result.success(true)
+        }
+        .onErrorReturn { Result.failure(it) }
+        .subscribe()
```

Comment:
I feel like I'm missing something here, I'm guessing maybe I should be returning this result, rather than subscribing here? But I don't fully understand what would be consuming the result.

As is this correctly launches the permissions dialog, but it seems that the onPermissionRequestResult isn't called afterwards, I'm guessing because there are no longer any subscribers?

@gino-m do you know what I should be doing with the result here  and how I can make sure the onPermissionRequestResult is called?

Reply:
Right.. if you turn this into a suspend function you can call `await()` instead of subscribing.

Reply:
You could also get rid of the chained calls to `andThen()`, instead calling `await()` on each step in the flow.

Reply:
Resolved after switching to a suspend fun